tests/kafkatest/tests/streams/streams_upgrade_test.py

```diff
@@ -0,0 +1,115 @@
+from ducktape.mark.resource import cluster
+from ducktape.tests.test import Test
+from ducktape.mark import parametrize
+from kafkatest.services.kafka import KafkaService
+from kafkatest.services.zookeeper import ZookeeperService
+from kafkatest.services.streams import StreamsSmokeTestDriverService, StreamsSmokeTestJobRunnerService
+from kafkatest.version import LATEST_0_10_1, LATEST_0_10_2, DEV_BRANCH, KafkaVersion
+import time
+
+
+class StreamsUpgradeTest(Test):
+    """
+    Tests rolling upgrades and downgrades of the Kafka Streams library.
+    """
+
+    def __init__(self, test_context):
+        super(StreamsUpgradeTest, self).__init__(test_context)
+        self.replication = 3
+        self.partitions = 1
+        self.isr = 2
+        self.topics = {
+            'echo' : { 'partitions': self.partitions, 'replication-factor': self.replication,
+                       'configs': {"min.insync.replicas": self.isr}},
+            'data' : { 'partitions': self.partitions, 'replication-factor': self.replication,
+                       'configs': {"min.insync.replicas": self.isr} },
+            'min' : { 'partitions': self.partitions, 'replication-factor': self.replication,
+                      'configs': {"min.insync.replicas": self.isr} },
+            'max' : { 'partitions': self.partitions, 'replication-factor': self.replication,
+                      'configs': {"min.insync.replicas": self.isr} },
+            'sum' : { 'partitions': self.partitions, 'replication-factor': self.replication,
+                      'configs': {"min.insync.replicas": self.isr} },
+            'dif' : { 'partitions': self.partitions, 'replication-factor': self.replication,
+                      'configs': {"min.insync.replicas": self.isr} },
+            'cnt' : { 'partitions': self.partitions, 'replication-factor': self.replication,
+                      'configs': {"min.insync.replicas": self.isr} },
+            'avg' : { 'partitions': self.partitions, 'replication-factor': self.replication,
+                      'configs': {"min.insync.replicas": self.isr} },
+            'wcnt' : { 'partitions': self.partitions, 'replication-factor': self.replication,
+                       'configs': {"min.insync.replicas": self.isr} },
+            'tagg' : { 'partitions': self.partitions, 'replication-factor': self.replication,
+                       'configs': {"min.insync.replicas": self.isr} }
+        }
+        
+
+    def perform_upgrade(self, to_version):
+        self.logger.info("First pass bounce - rolling upgrade")
+
+        # get the node running the streams app
+        node = self.processor1.node
+        self.processor1.stop()
+
+        # change it's version. This will automatically make it pick up a different
+        # JAR when it starts again
+        node.version = KafkaVersion(to_version)
+        self.processor1.start()
+        
+    @cluster(num_nodes=6)
+    @parametrize(from_version=str(LATEST_0_10_1), to_version=str(DEV_BRANCH))
```

Comment:
We will once it's released since we don't have a package for it on S3 yet. I expect we'll add to this test over time, just like we do in the Kafka upgrade test.